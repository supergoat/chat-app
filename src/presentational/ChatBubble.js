// @flow
import React, { Component } from 'react';
import type { NavigationScreenProp } from 'react-navigation';
import {
  Text,
  View
} from 'react-native';
import CAStyleSheet from "../common/CAStyleSheet";
import CAColors from "../common/CAColors";
import { formatTime } from '../common/CAHelpers';


/** ============================================================================
<ChatBubble />
--------------------------------------------------------------------------------
Displays a message
============================================================================= */
type Props = {
  isAlignedLeft: boolean,
  message: {
    id: string,
    text: string,
    createdAt: Date,
    createdBy: string
  }
};
class ChatBubble extends Component<Props> {
  static __cards__: Function

  render() {
    const { message, isAlignedLeft } = this.props;
    return (
      <View style={[
        styles.messageCard,
        { backgroundColor: isAlignedLeft ? CAColors.alabaster : CAColors.orinoco },
        { alignSelf: isAlignedLeft ? 'flex-start': 'flex-end' },
        this.props.style
      ]}>
        <Text style={styles.text}>
          {message.text}
          {/* HACK:  Add this to position the timeStamp correctly */}
          <Text style={styles.hide}>{'xxxxxx'}</Text>
        </Text>
        <Text style={styles.timeStamp}>{formatTime(message.createdAt)}</Text>
      </View>
    )
  }
}

/* StyleSheet
============================================================================= */
const styles = CAStyleSheet.create({
  messageCard: {
    paddingVertical: 5,
    paddingHorizontal: 7,
    marginBottom: 3,
    marginHorizontal: 20,
    borderRadius: 6,
    maxWidth: '78%',
    ios: {
      shadowColor: CAColors.osloGrey,
      shadowOffset: {
        width: 0,
        height: 0.5
      },
      shadowOpacity: 1,
      shadowRadius: 0,
    },
    android: {
      elevation: 1
    }
  },
  text: {
    lineHeight: 19,
    fontSize: 15,
    color: CAColors.black
  },
  hide: {
    opacity: 0,
    color: 'transparent'
  },
  timeStamp: {
    position: 'absolute',
    right: 8,
    bottom: 3,
    color: CAColors.osloGrey,
    fontSize: 11
  }
});

/* Playground Cards - the different states of ChatBubble
============================================================================= */
// ChatBubble.__cards__ = define => {
//   define("Simple", () => {
//     return <ChatBubble />
//   });
// };

/* Export
============================================================================= */
export default ChatBubble;
