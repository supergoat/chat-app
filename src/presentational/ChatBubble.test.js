// @flow
import React from 'react';
import { Text } from 'react-native';
import ChatBubble from './ChatBubble';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';


describe('ChatBubble', () => {
  // // Custom constant date, so our snapshots don't fail when the date changes
  // const customDate = new Date(1, 1, 3, 9, 5);
  // const RealDate = Date;
  //
  // beforeEach(() => {
  //   // Mock Date to always return our custom date
  //   global.Date = class extends RealDate {
  //     constructor() {
  //       return new RealDate(customDate);
  //     }
  //   };
  // });
  //
  // afterEach(() => {
  //   global.Date = RealDate;
  // });

  it('renders correctly', () => {
    const message = {
      id: '1',
      text: 'Hello!',
      createdAt: new Date(1, 1, 3, 9, 5),
      createdBy: 'Panayiotis'
    };

    const tree = renderer.create(
      <ChatBubble
        message={message}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when isAlignedLeft is set to to true', () => {
    const message = {
      id: '1',
      text: 'Hello!',
      createdAt: new Date(1, 1, 3, 9, 5),
      createdBy: 'Panayiotis'
    };

    const tree = renderer.create(
      <ChatBubble
        isAlignedLeft={true}
        message={message}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('has a hidden text field used to correctly align the timestamp', () => {
    const message = {
      id: '1',
      text: 'Hello!',
      createdAt: new Date(1, 1, 3, 9, 5),
      createdBy: 'Panayiotis'
    };

    const component = shallow(
      <ChatBubble
        isAlignedLeft={true}
        message={message}
      />
    )

    expect(component.find(Text).at(1).exists()).toBe(true);
    expect(component.find(Text).at(1).prop('children')).toEqual('xxxxxx');
    expect(component.find(Text).at(1).prop('style')).toEqual({"color": "transparent", "opacity": 0});
  });
});
