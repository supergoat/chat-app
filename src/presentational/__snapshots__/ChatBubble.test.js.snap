// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatBubble renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 6,
        "marginBottom": 3,
        "marginHorizontal": 20,
        "maxWidth": "78%",
        "paddingHorizontal": 7,
        "paddingVertical": 5,
        "shadowColor": "#7e8890",
        "shadowOffset": Object {
          "height": 0.5,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 0,
      },
      Object {
        "backgroundColor": "#DFF8C8",
      },
      Object {
        "alignSelf": "flex-end",
      },
      undefined,
    ]
  }
>
  <Text
    style={
      Object {
        "color": "#000000",
        "fontSize": 15,
        "lineHeight": 19,
      }
    }
  >
    Hello!
    <Text
      style={
        Object {
          "color": "transparent",
          "opacity": 0,
        }
      }
    >
      xxxxxx
    </Text>
  </Text>
  <Text
    style={
      Object {
        "bottom": 3,
        "color": "#7e8890",
        "fontSize": 11,
        "position": "absolute",
        "right": 8,
      }
    }
  >
    09:05
  </Text>
</View>
`;

exports[`ChatBubble renders correctly when isAlignedLeft is set to to true 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 6,
        "marginBottom": 3,
        "marginHorizontal": 20,
        "maxWidth": "78%",
        "paddingHorizontal": 7,
        "paddingVertical": 5,
        "shadowColor": "#7e8890",
        "shadowOffset": Object {
          "height": 0.5,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 0,
      },
      Object {
        "backgroundColor": "#F9F9F9",
      },
      Object {
        "alignSelf": "flex-start",
      },
      undefined,
    ]
  }
>
  <Text
    style={
      Object {
        "color": "#000000",
        "fontSize": 15,
        "lineHeight": 19,
      }
    }
  >
    Hello!
    <Text
      style={
        Object {
          "color": "transparent",
          "opacity": 0,
        }
      }
    >
      xxxxxx
    </Text>
  </Text>
  <Text
    style={
      Object {
        "bottom": 3,
        "color": "#7e8890",
        "fontSize": 11,
        "position": "absolute",
        "right": 8,
      }
    }
  >
    09:05
  </Text>
</View>
`;
